### PR TITLE
Refactors `promotions.html.erb`

### PR DESCRIPTION
**Pull request by sarthaktexas**
## Summary of the problem

Refactors it.

## Describe your changes

Creates a partial and uses that.

TODO:
- this is to show elibgible perks automatially - gary separrte pr
- convert to tailwind

No visual changes.

### CODE REVIEW

File: app/views/events/_promotion.html.erb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+        </h3>
+      <%= link_to local_assigns[:button_name] || "Lorem Ipsum",
+        @perks_available ? local_assigns[:button_link] || "" : root_path,
+        class: "btn #{local_assigns[:button_class]} mr1 #{local_assigns[:conditional_disable] ? "disabled" : nil} tooltipped tooltipped--w",
+        "aria-label" => local_assigns[:button_label] || "",
+        target: :_blank %>
+    </div>
+    <%= yield %>
+  </div>
+<% end %>
```
Comment (garyhtou) [inline app/views/events/_promotion.html.erb]:
```suggestion
<% end %>

```

File: app/views/events/promotions.html.erb [hunk 2]
```diff
@@ -8,53 +8,49 @@
...
-        @perks_available ? "https://hackclub.com/hackathons/grant/" : root_path,
-        class: "btn mr1 #{'disabled' unless @perks_available} tooltipped tooltipped--w",
-        "aria-label" => "Get notified when grant applications are open.",
-        target: :_blank %>
-    </div>
-
-    <p>
+<%# Hackathon Grant %>
+<%= render :layout => 'promotion', :locals => {
+  :hidden => !(@event.hackathon? || @event.high_school_hackathon?) || !(@event.country_us? || @event.country == “CA”),
```
Comment (garyhtou) [inline app/views/events/promotions.html.erb]:
```suggestion
  :hidden => !(@event.hackathon? || @event.high_school_hackathon?) || !(@event.country_us? || @event.country == “CA”),
  :conditional_disable => @event.demo_mode?,
```

File: app/views/events/_promotion.html.erb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<% unless local_assigns[:hidden] %>
+<% !@perks_available ? local_assigns[:conditional_disable] = true : nil%>
```
Comment (garyhtou) [inline app/views/events/_promotion.html.erb]: we can drop this line if we add a `conditional_disable` to each partial render.

Comment (garyhtou) [inline app/views/events/_promotion.html.erb]: rename `conditional_disable` -> `disabled`

File: app/views/events/_promotion.html.erb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  <div class="card pb0 mb3 <%= local_assigns[:conditional_disable] ? "muted" : nil %>" id="<%= local_assigns[:promo_id] %>" style="<%= local_assigns[:promo_style] %>">
```
Comment (garyhtou) [inline app/views/events/_promotion.html.erb]:
```suggestion
  <div class="card pb0 mb3 <%= "muted" if local_assigns[:conditional_disable] %>" id="<%= local_assigns[:promo_id] %>" style="<%= local_assigns[:promo_style] %>">
```

Comment (garyhtou) [inline app/views/events/_promotion.html.erb]:
```suggestion
  <div class="card pb0 mb3 <%= local_assigns[:conditional_disable] ? "muted" : nil %>" id="<%= promo_id %>" style="<%= local_assigns[:promo_style] %>">
```

File: app/views/events/_promotion.html.erb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+<% unless local_assigns[:hidden] %>
```
Comment (garyhtou) [inline app/views/events/_promotion.html.erb]: Add a magic comment for strict locals

https://blog.kiprosh.com/allow-template-to-set-strict-locals/

File: app/views/events/_promotion.html.erb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    <div class="card__banner card__banner--top flex justify-between items-center promo-<%= local_assigns[:promo_id] %>">
+      <%= local_assigns[:title_image] %>
+      <h3 class="h1 p2 my0 flex flex-column items-start <%= local_assigns[:header_class] %>">
+          <% if local_assigns[:new] %>
+            <span class="badge bg-success m0 mb1">New!</span>
+          <% end %>
+          <%= local_assigns[:title] %>
+        </h3>
+      <%= link_to local_assigns[:button_name] || "Lorem Ipsum",
+        @perks_available ? local_assigns[:button_link] || "" : root_path,
```
Comment (garyhtou) [inline app/views/events/_promotion.html.erb]: Let's drop usage of `@perks_available`

File: app/views/events/_promotion.html.erb [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+        class: "btn #{local_assigns[:button_class]} mr1 #{local_assigns[:conditional_disable] ? "disabled" : nil} tooltipped tooltipped--w",
```
Comment (garyhtou) [inline app/views/events/_promotion.html.erb]:
```suggestion
        class: "btn #{local_assigns[:button_class]} mr1 #{"disabled" if local_assigns[:conditional_disable]} tooltipped tooltipped--w",
```

File: app/views/events/promotions.html.erb [hunk 8]
```diff
@@ -8,53 +8,49 @@
...
-      <%= link_to "Apply",
-        @perks_available ? "https://hackclub.com/hackathons/grant/" : root_path,
-        class: "btn mr1 #{'disabled' unless @perks_available} tooltipped tooltipped--w",
-        "aria-label" => "Get notified when grant applications are open.",
-        target: :_blank %>
-    </div>
-
-    <p>
+<%# Hackathon Grant %>
+<%= render :layout => 'promotion', :locals => {
```
Comment (garyhtou) [inline app/views/events/promotions.html.erb]: Currently, this is written using the hash rocket syntax. let's switch to the JSON syntax

File: app/views/events/promotions.html.erb [hunk 9]
```diff
@@ -8,53 +8,49 @@
...
+  :button_class => "bg-success",
+  :button_link => "https://hackclub.com/hackathons/grant/",
+  :button_label => "Get notified when grant applications are open.",
+} do %>
+  <%= link_to "1Password", "https://1password.com" %> is providing $100 USD
     credit to all organizations on HCB that can be claimed once per
     organizations. 1Password is a password manager that allows you to securely
     keep and share account logins and credentials with your teammates; we use it
     ourselves at Hack Club HQ!
   </p>
```
Comment (garyhtou) [inline app/views/events/promotions.html.erb]: this is missing an opening tag

File: app/views/events/promotions.html.erb [hunk 10]
```diff
@@ -8,53 +8,49 @@
...
-      class: "btn mr1 #{'disabled' unless @perks_available}",
-      target: :_blank %>
-  </div>
+<% end %>
 
+<%# Event Helper %>
+<%= render :layout => 'promotion', :locals => {
+  :hidden => true,
+  :promo_id => "event-helper",
+  :title_image => (image_tag "https://cloud-fz66hmje0-hack-club-bot.vercel.app/0theeventhelperbutbetter.png", width: 60),
```
Comment (garyhtou) [inline app/views/events/promotions.html.erb]:
```suggestion
  :title_image => image_tag("https://cloud-fz66hmje0-hack-club-bot.vercel.app/0theeventhelperbutbetter.png", width: 60),
```

Comment (garyhtou) [reply app/views/events/promotions.html.erb]: Same thing for the other ones

File: app/views/events/promotions.html.erb [hunk 11]
```diff
@@ -88,38 +84,17 @@
...
-    <%= image_tag "https://cloud-q7i8htyv7-hack-club-bot.vercel.app/0pvsa-logo-2x.png", width: 60 %>
-    <%= link_to "Purchase",
-      @perks_available ? fillout_form("dyHShKvTL1us", { "Event Name" => @event.name, "Bank URL" => event_url(@event), "Submitter Email" => current_user.email }, prefix: "prefill_") : root_path,
-      class: "btn mr1 #{'disabled' unless @perks_available}",
-      target: :_blank %>
-  </div>
-
+<%# PVSA %>
+<%= render :layout => 'promotion', :locals => {
+  :hidden => true,
```
Comment (garyhtou) [inline app/views/events/promotions.html.erb]: do we want this permanently hidden?

File: app/views/events/_promotion.html.erb (hunk 6 above)
Comment (garyhtou) [reply app/views/events/_promotion.html.erb]: If disabled, then make the link the root_path. This way, we don't need to add a similar condition to every single partial render

File: app/views/events/promotions.html.erb [hunk 12]
```diff
@@ -160,25 +135,20 @@
...
-  <p class="mb0">
+<%# Domains %>
+<%= render :layout => 'promotion', :locals => {
+  :hidden => true,
+  :conditional_disable => !@event.g_suites.none?,
+  :promo_id => "domains",
+  :title => "Domains",
+  :button_name => "Apply",
+  :button_class => "bg-success",
+  :button_link => @perks_available ? fillout_form("pUUaNHEiT1us", { "hcbprojname" => @event.name, "hcburl" => event_url(@event), "name" => current_user.name, "email" => current_user.email }) : root_path,
```
Comment (garyhtou) [inline app/views/events/promotions.html.erb]: Drop the condition for @perks_available

Comment (garyhtou) [reply app/views/events/promotions.html.erb]: Same for others

**Comment by garyhtou** [comment]
FYI, merge conflicts from https://github.com/hackclub/hcb/pull/6299

**Comment by sarthaktexas** [comment]
Closing. This is pretty hard because promotions are so different now (StickerNinja has a copy-pastable code).

**Pull request closed without merging**